Class/CDC/ECM/usbh_ecm.c

```diff
@@ -0,0 +1,477 @@
+*
+* Return(s)   : Pointer to CDC ECM device.
+*
+* Note(s)     : (1) ECM 1.2 Section 3.3, select alternate interface for network traffic. Default interface does not
+*                   have any endpoints and does not support data transfers.
+*********************************************************************************************************
+*/
+
+USBH_CDC_ECM_DEV  *USBH_CDC_ECM_Add (USBH_CDC_DEV  *p_cdc_dev,
+                                     USBH_ERR      *p_err)
+{
+    USBH_CDC_ECM_DEV   *p_cdc_ecm_dev;
+    USBH_IF            *p_cif;
+    USBH_CDC_ECM_DESC   ecm_desc;
+    LIB_ERR             err_lib;
+
+
+    if (p_cdc_dev == (USBH_CDC_DEV *)0) {
+       *p_err = USBH_ERR_INVALID_ARG;
+        return ((USBH_CDC_ECM_DEV *)0);
+    }
+                                                                                /* Allocate CDC ECM dev from mem pool.  */
+    p_cdc_ecm_dev = (USBH_CDC_ECM_DEV *)Mem_PoolBlkGet(&USBH_CDC_ECM_DevPool,
+                                                        sizeof(USBH_CDC_ECM_DEV),
+                                                       &err_lib);
+    if (err_lib != LIB_MEM_ERR_NONE) {
+       *p_err = USBH_ERR_ALLOC;
+        return ((USBH_CDC_ECM_DEV *)0);
+    }
+
+    p_cdc_ecm_dev->CDC_DevPtr = p_cdc_dev;
+
+    p_cif = USBH_CDC_CommIF_Get(p_cdc_ecm_dev->CDC_DevPtr);
+   *p_err = USBH_CDC_ECM_DescParse(&ecm_desc, p_cif);                           /* Get ECM desc from IF.                */
+    if (*p_err != USBH_ERR_NONE) {
+        goto end_free_return_empty;
+    }
+
+                                                                                /* Get MAC address from string desc.    */
+    USBH_StrGet(p_cdc_dev->DevPtr,ecm_desc.iMACAddress, 0, p_cdc_ecm_dev->MAC_Addr,
+                sizeof(p_cdc_ecm_dev->MAC_Addr), p_err);
+    if (*p_err != USBH_ERR_NONE) {
+        goto end_free_return_empty;
+    }
+
+                                                                                /* Save parameters.                     */
+    p_cdc_ecm_dev->Parameters.EthernetStatistics = ecm_desc.bmEthernetStatistics;
+    p_cdc_ecm_dev->Parameters.MaxSegmentSize     = ecm_desc.wMaxSegmentSize;
+    p_cdc_ecm_dev->Parameters.NumberMCFilters    = ecm_desc.wNumberMCFilters;
+    p_cdc_ecm_dev->Parameters.NumberPowerFilters = ecm_desc.bNumberPowerFilters;
+
+                                                                                /* Select CDC ECM configuration.        */
+   *p_err = USBH_CfgSet(p_cdc_dev->DevPtr, p_cdc_dev->Cfg_Nbr);
+    if (*p_err != USBH_ERR_NONE) {
+        goto end_free_return_empty;
+    }
+
+                                                                                /* See Note #1.                         */
+    USBH_SET_IF(p_cdc_dev->DevPtr, p_cdc_dev->DIC_IF_Nbr, p_cdc_dev->DIC_IF_Ptr->AltIxSel, p_err);
+    if (*p_err != USBH_ERR_NONE) {
+        goto end_free_return_empty;
+    }
+
+    return (p_cdc_ecm_dev);
+
+end_free_return_empty:
```

Comment:
remove line 262-266. The memory pool block free is done during the `USBH_CDC_ECM_Remove()`

Reply:
I am not sure to follow. In case of error `USBH_CDC_ECM_Add` will return `0`, which will loose the pointer we need to free, no? That's why we have that logic in `USBH_CDC_ECM_Add` and cannot rely on `USBH_CDC_ECM_Remove`. Or am I missing something?

Reply:
I see what you mean and yes the pointer will be lost. Your point is valid  :D